### ISSUE TITLE
Feature Request - Enhanced Space Deletion - To delete OpenShift resources and remove need for Env Reset

### ISSUE THREAD

**Issue by ldimaggi** [openshiftio/openshift.io#4657]
As of today, human users and automated tests do not have a reliable way to remove a space, and all of its related OpenShift resources. The users/tests are able to work-around this by performing an environment reset, but this is problematic as the reset intermittently results in some resources not restarting. For example, issue #4598 describes an intermittent problem where Jenkins is not accessible after an environment reset.

In addition, the environment reset is a does not give the user fine-grained control over its actions. It is a destructive function that intentionally removes ALL user resources and restarts the user's pods. If a user wants to remove all resources associated with a single space, but not all resources associated with other spaces, then he/she cannot use the environment reset feature.

For automated tests, the environment reset function is also not reliable as after the reset, there are intermittent problems with Jenkins not starting. If a human user encounters this problem, the workaround is to reset the environment a second time. For automated tests, its less practical to reset the environment a variable number of times. 

A reliable space deletion function that also removes all associated OpenShift resources is preferable for both human users and automated tests. 

Related issues:
#4121 
#4598

**Comment by christianvogt** [comment]
This is also important for devs to run e2e tests with their own accounts. We encourage devs to run e2e tests and also fix them up. However if they need to always reset their env it's a big ask depending on the setup they already have. It would best to have a cleaner solution.

**Comment by pmacik** [comment]
> A reliable space deletion function that also removes all associated OpenShift resources is preferable for both human users and automated tests.

In addition, all the corresponding Che workspaces from the space's codebases should be removed as well.

**Comment by ldimaggi** [comment]
Will the fix to this issue:
https://github.com/fabric8-services/fabric8-tenant/issues/710

Also solve our env reset issues?
@MatousJobanek

**Comment by ljelinkova** [comment]
If I understand it correctly, the reset of environment calls delete on all spaces. That means that if delete space does not work properly, the reset of environment does not work either. 

That is why it is so annoying to have an issue like this where delete space does not delete Che workspaces.

https://github.com/openshiftio/openshift.io/issues/4559

It means we need to manually delete all che workspaces for all our testing accounts periodically (every few weeks) otherwise the Che tests stop to work properly. 

Another issue is that neither work items are deleted

https://github.com/openshiftio/openshift.io/issues/4382

BTW, I do not think this is a feature request, the space deletion is broken and should be fixed.

**Comment by ppitonak** [comment]
Pipelines are not removed on space delete, too https://github.com/openshiftio/openshift.io/issues/3802

**Comment by joshuawilson** [comment]
@ppitonak do you agree with the change to the labels and setting of severity?

**Comment by michaelkleinhenz** [comment]
Delete spaces seems to be broken or non complete in WIT. @kwk is already working on fixing this. As far as I collected from the issues linked here, deleting a space should:

 1. delete the Che workspaces via che-starter (#4559).
 2. delete all work items (#4382) and all other WIT resources that are connected to the space.

Is this list complete (minus the things already being done, see below)? The current implementation does the following as far as I can see:

 1. deletes the codebases.
 2. deletes deployments.
 3. deletes the space iteself.

**Comment by ppitonak** [comment]
@joshuawilson yes, thanks for setting it

@michaelkleinhenz #2352 is talking about idling Jenkins jobs that are waiting on approval, it has nothing to do with #3802

**Comment by michaelkleinhenz** [comment]
> @michaelkleinhenz #2352 is talking about idling Jenkins jobs that are waiting on approval, it has nothing to do with #3802

got it. Wanted to link to PR 2352, not issue 2352 :)

**Comment by ldimaggi** [comment]
+1 on Lucia's comment - changed the issue title to reflect that his is a bug and not a feature request. Thx!

**Comment by alexeykazakov** [comment]
I've removed team/platform because I don't see what else platform services should remove during space deletion. Please let me know if I missed something. I added team/wit because it needs to remove codebases, workspaces, work items etc.
